File: AutoMix Software/AutoMix UI/Presenter.cpp
// Copyright (C) 2016-2017 LesProjecteurs - All Rights Reserved
// Maxime STEVENOT, Guillaume HANNES, Jordan ERNULT, Louis CARLIER, Pierre GABON
// 
// This file is part of AutoMix Software.
// 
// Unauthorized copying of this file, via any medium is strictly prohibited.
// You should have received a copy of the License along with this program.

#include "Presenter.h"

namespace AutoMixUI {
	using namespace System;
	using namespace System::Collections;
	using namespace AutoMixAI;

	Presenter::Presenter()
	{
		_views = gcnew Generic::List<ViewWithTrackCollection^>();
		_trackCollection = gcnew TrackCollection();
		_dataExtractionEngine = gcnew AudioDataExtractionProxy();

		_sortAlgorithm = gcnew SimulatedAnnealingSortAlgorithm(gcnew SimpleDistance());
	}

	Presenter::Presenter(ViewWithTrackCollection^ view) : Presenter()
	{
		_views->Add(view);
	}

	void Presenter::notify()
	{
		for each (auto view in _views)
		{
			view->update(_trackCollection);
		}
	}
	bool Presenter::IsTrackCollectionFilled::get()
	{
		if (!_trackCollection)
		{
			return false;
		}
		return _trackCollection->Count >= 1;
	}
	TrackCollection^ Presenter::sortTrackCollectionWithGeneticAlgorithm(ComponentModel::BackgroundWorker^ bw)
	{
		if (_trackCollection->Count > 1)
		{
			_trackCollection = _sortAlgorithm->sort(bw, _trackCollection);
		}
		return _trackCollection;
	}

	TrackCollection^ Presenter::loadTracks(ComponentModel::BackgroundWorker^ bw, array<String^>^ fileEntries)
	{
		IEnumerator^ filesEnum = fileEntries->GetEnumerator();
		TrackCollection^ collection = gcnew TrackCollection();
		int count = 1;

		while (filesEnum->MoveNext() && !bw->CancellationPending)
		{
			String^ filePath = safe_cast<String^>(filesEnum->Current);

			if (IO::Directory::Exists(filePath))
			{
				IEnumerator^ filesInDirectoryEnum = IO::Directory::GetFiles(filePath)->GetEnumerator();

				while (filesInDirectoryEnum->MoveNext() && !bw->CancellationPending)
				{
					String^ filePathInDirectory = safe_cast<String^>(filesInDirectoryEnum->Current);

					if (Utils::getExtension(filePathInDirectory)->Contains("mp3"))
					{
						Track^ track = gcnew Track(filePathInDirectory);
						collection->safeAdd(track);
					}
				}
			}

			else if (Utils::getExtension(filePath)->Contains("mp3"))
			{
				Track^ track = gcnew Track(filePath);
				collection->safeAdd(track);
			}
			bw->ReportProgress((int)500 * count++ / fileEntries->Length);
		}

		_trackCollection->concat(collection);
		_trackCollection->sortByName();
		bw->ReportProgress(500);

		_dataExtractionEngine->extractData(bw, collection);
		retrieveControlOnCollection();

		_trackCollection->purge();
		return _trackCollection;
	}

	void Presenter::removeTracks(Generic::List<String^>^ selection)
	{
		for each (String^ trackName in selection)
		{
			_trackCollection->Remove(trackName);
		}
		notify();
	}

	void Presenter::moveTrack(int newIndex, String^ trackName)
	{
		if (newIndex < 0 && newIndex >= _trackCollection->Count)
		{
			notify();
			return;
		}

		Track^ track = _trackCollection->search(trackName);
		_trackCollection->Remove(track);
		_trackCollection->Insert(newIndex, track);
	}

	void Presenter::exportTrackList(System::ComponentModel::BackgroundWorker^ bw, String^ destinationFile)
	{

		if (_trackCollection->Count >= 1)
		{
			//createTransition(); BUGGED
			_trackCollection->exportToMP3(bw, destinationFile);
		}
	}

	void Presenter::clearDataBase()
	{
		DataBase^ db = gcnew DataBase();
		db->clear();
	}

	void Presenter::playMix(System::String ^ fileName)
	{
		_mp3Playing = gcnew AutoMixDataManagement::MP3Playing(fileName);
		_mp3Playing->play();
	}

	void Presenter::resumeMix()
	{
		_mp3Playing->play();
	}

	void Presenter::pauseMix()
	{
		_mp3Playing->pause();
	}

	void Presenter::stopMix()
	{
		_mp3Playing->stop();
	}

	void Presenter::seek(double seconds)
	{
		_mp3Playing->seek(seconds);
	}

	__int64 Presenter::getPosition()
	{
		return _mp3Playing->getPosition();
	}

	__int64 Presenter::getLength()
	{
		return _mp3Playing->getLength();
	}

	void Presenter::retrieveControlOnCollection()
	{
		_trackCollection = TrackCollection::CopyFrom(_trackCollection);
	}

	void Presenter::createTransition()
	{
		Transition^ transition = gcnew Transition(_trackCollection);
		transition->makeTransition();
	}
}

File: Automix UI/Presenter.cs
﻿// Copyright (C) 2016 - 2018 LesProjecteurs
// This file is part of Automix Software licensed under MIT License.

using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.IO;
using System.Reflection;
using Automix_AI.Distances;
using Automix_AI.Sort_Algorithms;
using Automix_Data_Management;
using Automix_Data_Management.Audio_Playing;
using Automix_Data_Management.Extraction;
using Automix_Data_Management.Model;
using Automix_Data_Management.Storage;
using log4net;

namespace Automix_UI
{
    internal class Presenter
    {
        private readonly List<IViewWithTrackCollection> _views;
        private TrackCollection _trackCollection;
        private IProfileDistance _sortProfile;

        private readonly IAudioDataExtraction _dataExtractionEngine;
        private AbstractSortAlgorithm _sortAlgorithm;
        private Mp3Player _mp3Player;

        private static readonly ILog log = LogManager.GetLogger(MethodBase.GetCurrentMethod().DeclaringType);

        public Presenter()
        {
            _views = new List<IViewWithTrackCollection>();
            _trackCollection = new TrackCollection();
            _dataExtractionEngine = new AudioDataExtractionProxy();
            _sortProfile = new BasicProfile();

            _sortAlgorithm = new SimulatedAnnealingSortAlgorithm(new SimpleTracksDistance(_sortProfile));
        }

        public Presenter(IViewWithTrackCollection view) : this()
        {
            _views.Add(view);
        }

        public bool IsTrackCollectionFilled => _trackCollection?.Count >= 1;

        public void ExportPlaylistInTextFile(string outputFilePath) => _trackCollection.ExportToText(outputFilePath);


        public void ClearDataBase()
        {
            var db = new DataBase();
            db.Clear();
        }

        public void MoveTrack(int dropIndex, string trackName)
        {
            if (dropIndex < 0 && dropIndex >= _trackCollection.Count)
            {
                Notify();
                return;
            }

            var track = _trackCollection.Search(trackName);
            _trackCollection.Remove(track);
            _trackCollection.Insert(dropIndex, track);
        }

        public void RemoveTracks(List<string> selection)
        {
            foreach (var trackName in selection)
            {
                _trackCollection.Remove(trackName);
            }
            Notify();
        }

        public void LockTracks(List<string> selection)
        {
            foreach(var trackName in selection)
            {
                var track = _trackCollection.Search(trackName);
                if (track != null)
                {
                    track.IsFixed = true;
                }
            }
        }

        public void UnlockTracks(List<string> selection)
        {
            foreach(var trackName in selection)
            {
                var track = _trackCollection.Search(trackName);
                if(track != null)
                {
                    track.IsFixed = false;
                }
            }
        }

        public object LoadTracks(BackgroundWorker backgroundWorker, string[] fileNames)
        {
            var filesEnum = fileNames.GetEnumerator();
            var tempTrackCollection = new TrackCollection();
            var count = 1;

            while (filesEnum.MoveNext() && !backgroundWorker.CancellationPending)
            {
                var filePath = (string)(filesEnum.Current);

                if (Directory.Exists(filePath))
                {
                    var filesInDirectoryEnum = Directory.GetFiles(filePath).GetEnumerator();

                    while (filesInDirectoryEnum.MoveNext() && !backgroundWorker.CancellationPending)
                    {
                        var filePathInDirectory = (string)(filesInDirectoryEnum.Current);

                        if (Utils.GetExtension(filePathInDirectory).Contains("mp3"))
                        {
                            var track = new Track(filePathInDirectory);
                            tempTrackCollection.SafeAdd(track);
                        }
                    }
                }

                else if (Utils.GetExtension(filePath).Contains("mp3"))
                {
                    var track = new Track(filePath);
                    tempTrackCollection.SafeAdd(track);
                }
                backgroundWorker.ReportProgress(250 * count++ / fileNames.Length);
            }

            _trackCollection.Concat(tempTrackCollection);
            _trackCollection.SortByName();
            backgroundWorker.ReportProgress(250);

            _dataExtractionEngine.ExtractData(backgroundWorker, tempTrackCollection);
            RetrieveControlOnCollection();

            _trackCollection.Purge();
            return _trackCollection;
        }

        private void RetrieveControlOnCollection()
        {
            _trackCollection = TrackCollection.CopyFrom(_trackCollection);
        }

        public void Notify()
        {
            foreach (var view in _views)
            {
                view.Update(_trackCollection);
            }
        }

        public object SortTrackCollection(BackgroundWorker backgroundWorker)
        {
            if (_trackCollection.Count > 1)
            {
                _trackCollection = _sortAlgorithm.Sort(backgroundWorker, _trackCollection);
            }
            return _trackCollection;
        }

        public void ExportTrackList(BackgroundWorker backgroundWorker, string outputFilePath)
        {
            if (_trackCollection.Count >= 1)
            {
                _trackCollection.ExportToMp3(backgroundWorker, outputFilePath);
            }
        }

        public void Seek(double d) => _mp3Player.Seek(d);


        public void PlayMix(string fileName)
        {
            _mp3Player = new Mp3Player(fileName);
            _mp3Player.Play();
        }

        public void ResumeMix() => _mp3Player?.Play();

        public void PauseMix() => _mp3Player?.Pause();

        public void StopMix()
        {
            _mp3Player?.Stop();
            _mp3Player = null;
        }

        public long GetPlayerPosition()
        {
            //TODO : correct the error or at least catch the exception into the other classes
            try
            {
                return _mp3Player.GetPosition();
            }
            catch (NullReferenceException e)
            {
                log.Debug("Exception in presenter --> TODO correct error", e);
                throw e;
            }

        }

        public long GetPlayerLength()
        {
            return _mp3Player.GetLength();
        }

        public void ClearMusicList()
        {
            foreach (Track t in _trackCollection)
            {
                if (!_trackCollection.IsPresent(t))
                {
                    _trackCollection.Remove(t.Name);
                }
            }
            Notify();
        }

        public void UpdateSortAlgorithm(IProfileDistance profileChosen)
        {
            _sortProfile = profileChosen;
            _sortAlgorithm = new SimulatedAnnealingSortAlgorithm(new SimpleTracksDistance(profileChosen));
        }
    }
}